drivers/ipm/ipm_rpi_pico.c

```diff
@@ -0,0 +1,291 @@
+{
+#if DT_INST_NODE_HAS_PROP(0, memory_region)
+	/* Make sure sp is within available RAM, and pc is within the image. */
+	uint32_t sram_start = DT_REG_ADDR(DT_INST_PHANDLE(0, memory_region));
+	uint32_t sram_size = DT_REG_SIZE(DT_INST_PHANDLE(0, memory_region));
+
+	/* Stack pointer shall point within RAM assigned to the core. */
+	if (!address_in_range(cpu1_sp, sram_start, sram_size)) {
+		LOG_ERR("CPU1 stack pointer 0x%08x invalid.", cpu1_sp);
+		return -EINVAL;
+	}
+
+	LOG_DBG("CPU1 stack pointer: 0x%08x", cpu1_sp);
+#endif /* DT_INST_NODE_HAS_PROP(0, memory_region) */
+
+	/* Initial program counter shall point to code within the CPU1 partition. */
+	if (!address_in_range(cpu1_pc, DT_REG_ADDR(DT_INST_PHANDLE(0, execution_memory)),
+			      DT_REG_SIZE(DT_INST_PHANDLE(0, execution_memory)))) {
+		LOG_ERR("CPU1 reset pointer 0x%08x invalid.", cpu1_pc);
+		return -EINVAL;
+	}
+
+	LOG_DBG("CPU1 reset pointer: 0x%08x", cpu1_pc);
+	return 0;
+}
+#endif /* CONFIG_IPM_LOG_LEVEL >= LOG_LEVEL_DBG */
+#endif /* DT_INST_NODE_HAS_PROP(0, execution_memory) */
+
+static int rpi_pico_mailbox_init(const struct device *dev)
+{
+	/* Boot CPU1 if provided a memory region */
+#if DT_INST_NODE_HAS_PROP(0, execution_memory)
+#if DT_INST_NODE_HAS_PROP(0, image_source)
+	BUILD_ASSERT((DT_REG_SIZE(DT_INST_PHANDLE(0, execution_memory)) >=
+		      DT_REG_SIZE(DT_INST_PHANDLE(0, image_source))),
+		     "Image size must not exceed execution memory size");
+
+	BUILD_ASSERT(((DT_REG_ADDR(DT_INST_PHANDLE(0, execution_memory)) >=
+		       DT_REG_ADDR(DT_INST_PHANDLE(0, image_source)) +
+			       DT_REG_SIZE(DT_INST_PHANDLE(0, image_source))) ||
+		      (DT_REG_ADDR(DT_INST_PHANDLE(0, image_source)) >=
+		       DT_REG_ADDR(DT_INST_PHANDLE(0, execution_memory)) +
+			       DT_REG_SIZE(DT_INST_PHANDLE(0, execution_memory)))),
+		     "Image source memory must not overlap with execution memory");
+
+	void *src_mem = (void *)DT_REG_ADDR(DT_INST_PHANDLE(0, image_source));
+	void *exec_mem = (void *)DT_REG_ADDR(DT_INST_PHANDLE(0, execution_memory));
+
+	LOG_DBG("Copying image from %p to %p", src_mem, exec_mem);
+
+	memcpy(exec_mem, src_mem,
+	       MIN(DT_REG_SIZE(DT_INST_PHANDLE(0, execution_memory)),
+		   DT_REG_SIZE(DT_INST_PHANDLE(0, image_source))));
+#endif /* DT_INST_NODE_HAS_PROP(0, image_source) */
+
+	uint32_t cpu1_image_base = DT_REG_ADDR(DT_INST_PHANDLE(0, execution_memory));
+
+	uint32_t *cpu1_vector_table = (void *)cpu1_image_base;
+	uint32_t cpu1_sp = cpu1_vector_table[0];
+	uint32_t cpu1_pc = cpu1_vector_table[1];
+
+#if CONFIG_IPM_LOG_LEVEL >= LOG_LEVEL_DBG
+	if (rpi_pico_validate_image(cpu1_sp, cpu1_pc) != 0) {
+		return -EINVAL;
+	}
+	LOG_DBG("Launching CPU1 with vector table at 0x%p", (void *)cpu1_vector_table);
+#endif /* CONFIG_IPM_LOG_LEVEL >= LOG_LEVEL_DBG */
+
+	if (rpi_pico_reset_cpu1() != 0) {
+		LOG_ERR("CPU1 reset failed.");
+		return -EIO;
+	}
+
+	rpi_pico_boot_cpu1((uint32_t)cpu1_vector_table, cpu1_sp, cpu1_pc);
```

Comment:
It is probably not a good idea to include the CPU1 kicking to the mailbox driver. It would be better to consolidate it under the soc directory.

Also, the zephyr SMP code does not currently support rp2040/rp2350 configurations.
It is good to mark this feature as experimental.

Reply:
> > I think it's possible to move this whole block to the soc directory.
> 
> Where should I place the three memory regions then? I want to configure them in my app to set custom code partition for CPU1, and DT seems to be more logical for that than Kconfig.

Since it will be unique in the system, I think it's fine to use Kconfig, but if you use dts, I think it's better to create `dts/bindings/cpu/raspberrypi,rp2350.yaml` and define it there.

Reply:
Hm, okay. Some extra stuff: when #89758 lands, there could be an option to run two different core types at once, at least there is likely to be such a thing in HW. Would be nice to express that in configs as well. But okay, could live in Kconfig as well

Reply:
Great, working on that in my working tree for an AMP branch! Thanks!